File: dietideals-service/src/main/java/it/uninastudents/dietidealsservice/model/entity/Carta.java
package it.uninastudents.dietidealsservice.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import it.uninastudents.dietidealsservice.model.base.BaseEntity;
import jakarta.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.time.OffsetDateTime;

@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@Data
@Entity
@Table(name = "carta", uniqueConstraints ={
    @UniqueConstraint(columnNames = "numero")
})

public class Carta extends BaseEntity {

    @Column(nullable = false)
    private String numero;

    @Column(nullable = false)
    private String nomeTitolare;

    @Column(nullable = false, length = 3)
    private String codiceCvvCvc;

    @Column(nullable = false)
    private OffsetDateTime dataScadenza;

    @JsonIgnore
    @ManyToOne
    @JoinColumn(nullable = false)
    private Utente utente;

}

File: dietidealsapp/app/src/main/java/com/example/dietideals_app/model/Carta.kt
package com.example.dietideals_app.model

import java.util.UUID

class Carta(
    val id: UUID,
    val numero: String,
    val nomeTitolare: String,
    val codiceCvvCvc: String,
    val dataScadenza: String,
    val utente: Utente
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Carta) return false

        if (numero != other.numero) return false
        if (nomeTitolare != other.nomeTitolare) return false
        if (codiceCvvCvc != other.codiceCvvCvc) return false
        if (dataScadenza != other.dataScadenza) return false
        return utente == other.utente
    }

    override fun hashCode(): Int {
        var result = numero.hashCode()
        result = 31 * result + nomeTitolare.hashCode()
        result = 31 * result + codiceCvvCvc.hashCode()
        result = 31 * result + dataScadenza.hashCode()
        result = 31 * result + utente.hashCode()
        return result
    }
}